### ISSUE TITLE
Card dealing is potentially uneven

### ISSUE THREAD

**Issue by matteodelabre** [JanssenSimon/le10#17]
While playing, we noticed that the dealt hands are often uneven as compared to real-life deals, with most players often having 5 or 6 cards of the same suit.

The [current code for dealing cards](https://github.com/JanssenSimon/le10/blob/8eb5f21e34c38ebb93f976881dccb3c533857b00/game.js#L406) to players proceeds by iterating through the unshuffled deck card by card, assigning each card to a random player that does not yet have 10 cards in their hand (I’ll call this approach `dealSelectHand`).

```js
const dealSelectHand = () => {
    const arrs = [[],[],[],[]];
    suits.forEach((s) => {
        values.forEach((v) => {
            do {
                var selectedHand = Math.floor(Math.random() * arrs.length);
            } while (arrs[selectedHand].length >= 10);
            arrs[selectedHand].push(s+v);
        });
    });
    return arrs;
};
```

This is in contrast with a more “natural” approach, which would first shuffle the entire deck, then assign cards in the shuffled order to the four players (I’ll call this one `dealShuffleDeck`).

```js
const dealShuffleDeck = () => {
    const deck = [];
    suits.forEach(s => values.forEach(v => deck.push(s+v)));
    shuffle(deck);

    const arrs = [];
    for (let i = 0; i < 4; ++i) {
        arrs.push(deck.slice(i * 10, (i + 1) * 10));
    }
    return arrs;
};
```

To compare those two approaches, I ran some simulations on 1,000,000 deals using both methods, taking note of the suit distribution in each player’s hands. Here are the results:

`dealSelectHand`
```
4,3,2,1         23.39%
3,3,2,2         11.49%
5,3,2           8.60%
4,3,3           7.55%
5,2,2,1         7.51%
4,2,2,2         5.78%
4,4,2           5.51%
3,3,3,1         5.26%
5,3,1,1         4.91%
5,4,1           4.33%
6,2,1,1         3.13%
6,3,1           2.93%
4,4,1,1         2.86%
6,2,2           2.18%
7,2,1           1.59%
6,4             0.89%
5,5             0.58%
7,1,1,1         0.47%
7,3             0.44%
8,1,1           0.33%
8,2             0.19%
9,1             0.07%
10              0.00%
```

`dealShuffleDeck`
```
4,3,2,1         32.08%
3,3,2,2         20.63%
4,2,2,2         9.01%
3,3,3,1         8.16%
5,2,2,1         7.23%
5,3,1,1         4.29%
4,3,3           4.29%
5,3,2           3.87%
4,4,1,1         3.12%
4,4,2           2.81%
5,4,1           1.50%
6,2,1,1         1.34%
6,3,1           0.72%
6,2,2           0.61%
7,2,1           0.16%
6,4             0.06%
7,1,1,1         0.06%
5,5             0.05%
7,3             0.02%
8,1,1           0.01%
8,2             0.00%
9,1             0.00%
```

Notice that the `5,3,2` configuration has a 8.60% probability with the current method, and a 3.87% probability with the “natural” method.

**Comment by matteodelabre** [comment]
Here’s the code I used for the simulations:
```js
const suits = ["♡","♣","♢","♠"];
const values = ["5","6","7","8","9","10","J","Q","K","A"];

const dealSelectHand = () => {
    const arrs = [[],[],[],[]];
    suits.forEach((s) => {
        values.forEach((v) => {
            do {
                var selectedHand = Math.floor(Math.random() * arrs.length);
            } while (arrs[selectedHand].length >= 10);
            arrs[selectedHand].push(s+v);
        });
    });
    return arrs;
};

const shuffle = array => {
    // See <https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#The_modern_algorithm>
    for (let i = array.length - 1; i > 0; --i) {
        let j = Math.floor(Math.random() * (i + 1));
        [array[i], array[j]] = [array[j], array[i]];
    }
};

const dealShuffleDeck = () => {
    const deck = [];
    suits.forEach(s => values.forEach(v => deck.push(s+v)));
    shuffle(deck);

    const arrs = [];
    for (let i = 0; i < 4; ++i) {
        arrs.push(deck.slice(i * 10, (i + 1) * 10));
    }
    return arrs;
};

const computeStats = (counts, deal) => {
    for (let hand of deal) {
        const totals = {};
        for (let card of hand) {
            if (!totals.hasOwnProperty(card[0])) {
                totals[card[0]] = 1;
            } else {
                ++totals[card[0]];
            }
        }
        const shape = Object.values(totals).sort((a, b) => b - a).join();
        if (!counts.hasOwnProperty(shape)) {
            counts[shape] = 1;
        } else {
            ++counts[shape];
        }
    }
};

const evaluateMethod = (method, repeats) => {
    const counts = {};

    for (let i = 0; i < repeats; ++i) {
        computeStats(counts, method());
    }

    const tally = Object.entries(counts)
        .sort(([s1, a], [s2, b]) => b - a)
        .map(([shape, count]) => [shape, count / (repeats * 4)]);

    console.log(`${method.name}()`);

    for (const [shape, freq] of tally) {
        console.log(`${shape}\t\t${(100 * freq).toFixed(2)}%`);
    }

    console.log()
};

const repeats = 1000000;
evaluateMethod(dealSelectHand, repeats);
evaluateMethod(dealShuffleDeck, repeats);
```

**Comment by JanssenSimon** [comment]
J'ai écris une autre petite fonction de sort pour voir s'il était possible de condenser ça:
```js
const dealBruh = () => {
    const cards = values.flatMap(value => suits.flatMap(suit => suit+value));
    const s = cards.map(face => ({ face, sort: Math.random() })) //shuffle
                   .sort((a, b) => a.sort - b.sort)              //
                   .map(({ face }) => face);                     //
    const l = s.length/4;
    const arrs = [s.slice(0,l), s.slice(l,2*l), s.slice(2*l,3*l), s.slice(3*l)];
    return arrs;
};
```
Avec des distributions pareils au dealShuffleDeck()
```
dealBruh()
4,3,2,1		32.10%
3,3,2,2		20.65%
4,2,2,2		9.03%
3,3,3,1		8.19%
5,2,2,1		7.21%
4,3,3		4.29%
5,3,1,1		4.29%
5,3,2		3.83%
4,4,1,1		3.12%
4,4,2		2.80%
5,4,1		1.50%
6,2,1,1		1.34%
6,3,1		0.72%
6,2,2		0.60%
7,2,1		0.15%
6,4		0.06%
7,1,1,1		0.06%
5,5		0.04%
7,3		0.02%
8,1,1		0.01%
8,2		0.00%
9,1		0.00%
```
Mais avec une perfomance terrible (50s  pour runner evaluateMethod comparé à 15s pour les autres).

**Comment by JanssenSimon** [comment]
Resolved in commit fa76f4a179a2e791315c169c6f768dfc370dd587

Here's the new dealing code: https://github.com/JanssenSimon/le10/blob/fa76f4a179a2e791315c169c6f768dfc370dd587/game.js#L410-L433
